general/mem_manager.hpp

```diff
@@ -13,44 +13,463 @@
+
+        HOST < HOST_32 < HOST_64 < CUDA < CUDA_UVM. */
+MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
+
+/// Class used by MFEM to store pointers to host and/or device memory.
+/** The template class parameter, T, must be a plain-old-data (POD) type.
+
+    In many respects this class behaves like a pointer:
+    * When destroyed, a Memory object does NOT automatically delete any
+      allocated memory.
+    * Only the method Delete() will deallocate a Memory object.
+    * Other methods that modify the object (e.g. New(), Wrap(), etc) will simply
+      overwrite the old contents.
+    * One difference with a pointer is that a const Memory object does not allow
+      modification of the content (unlike e.g. a const pointer).
+
+    A Memory object stores up to two different pointers: one host pointer (with
+    MemoryType from MemoryClass::HOST) and one device pointer (currently one of
+    MemoryType::CUDA or MemoryTyep::CUDA_UVM).
+
+    A Memory object can hold (wrap) an externally allocated pointer with any
+    given MemoryType.
+
+    Access to the content of the Memory object can be requested with any given
+    MemoryClass through the methods ReadWriteAccess(), ReadAccess(), and
+    WriteAccess(). Requesting such access may result in additional (internally
+    handled) memory allocation and/or memory copy.
+    * When ReadWriteAccess() is called, the returned pointer becomes the only
+      valid pointer.
+    * When ReadAccess() is called, the returned pointer becomes valid, however
+      the other pointer (host or device) may remain valid as well.
+    * When WriteAccess() is called, the returned pointer becomes the only valid
+      pointer, however, unlike ReadWriteAccess(), no memory copy will be
+      performed.
+
+    The host memory (pointer from MemoryClass::HOST) can be accessed through the
+    inline methods: `operator[]()`, `operator*()`, the implicit conversion
+    functions `operator T*()`, `operator const T*()`, and the explicit
+    conversion template functions `operator U*()`,  `operator const U*()` (with
+    any suitable type U). In certain cases, using these methods may have
+    undefined behavior, e.g. if the host pointer is not currently valid. */
+template <typename T>
+class Memory
+{
+protected:
+   friend class MemoryManager;
+   friend void MemoryPrintFlags(unsigned flags);
+
+   enum FlagMask
+   {
+      REGISTERED    = 1,   ///< #h_ptr is registered with the MemoryManager
+      OWNS_HOST     = 2,   ///< The host pointer will be deleted by Delete()
+      OWNS_DEVICE   = 4,   ///< The device pointer will be deleted by Delete()
+      OWNS_INTERNAL = 8,   ///< Ownership flag for internal Memory data
+      VALID_HOST    = 16,  ///< Host pointer is valid
+      VALID_DEVICE  = 32,  ///< Device pointer is valid
+      ALIAS         = 64,
+      /// Execution flag reserved for use by MFEM, e.g. see Vector::UseDevice()
+      EXEC_FLAG     = 128
```

Comment:
The `EXEC_FLAG` could reflect better the use that we do latter in the code: most of the time it is to compute the `use_dev` logic with some code like: ```const bool use_dev = data.GetExecFlag() || v.data.GetExecFlag();```.
We also already tag most memory we want to use the device with `UseDevice();`, so just `GetUseDevice()` with `USE_DEVICE` might be a choice.

Reply:
Done